lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Script.hs

```diff
@@ -0,0 +1,176 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2021 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of shared script state using
+-- scheme specified in CIP-XXX Multi-signature Wallets.
+
+module Cardano.Wallet.Primitive.AddressDiscovery.Script
+    (
+      keyHashFromAccXPubIx
+    , constructAddressFromIx
+    , toNetworkTag
+    , replaceCosignersWithVerKeys
+    , liftPaymentAddress
+    , liftDelegationAddress
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( Cosigner (..)
+    , KeyHash (..)
+    , Script (..)
+    , ScriptHash (..)
+    , ScriptTemplate (..)
+    , toScriptHash
+    )
+import Cardano.Address.Style.Shelley
+    ( Credential (..)
+    , delegationAddress
+    , deriveMultisigForDelegationPublicKey
+    , deriveMultisigForPaymentPublicKey
+    , hashKey
+    , liftXPub
+    , mkNetworkDiscriminant
+    , paymentAddress
+    )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationType (..)
+    , Index (..)
+    , KeyFingerprint (..)
+    , NetworkDiscriminant (..)
+    , SoftDerivation
+    , WalletKey (..)
+    , deriveVerificationKey
+    , hashVerificationKey
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address (..) )
+import Data.Either.Combinators
+    ( fromRight', rightToMaybe )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Type.Equality
+    ( (:~:) (..), testEquality )
+import Type.Reflection
+    ( Typeable, typeRep )
+
+import qualified Cardano.Address as CA
+import qualified Cardano.Address.Style.Shelley as CA
+import qualified Data.Map.Strict as Map
+
+keyHashFromAccXPubIx
+    :: (SoftDerivation k, WalletKey k)
+    => k 'AccountK XPub
+    -> Index 'Soft 'ScriptK
+    -> KeyHash
+keyHashFromAccXPubIx accXPub ix =
+    hashVerificationKey $ deriveVerificationKey accXPub ix
+
+replaceCosignersWithVerKeys
+    :: CA.Role
+    -> ScriptTemplate
+    -> Index 'Soft 'ScriptK
+    -> Script KeyHash
+replaceCosignersWithVerKeys role (ScriptTemplate xpubs scriptTemplate) ix =
+    replaceCosigner scriptTemplate
+  where
+    replaceCosigner :: Script Cosigner -> Script KeyHash
```

Comment:
Side-note: this smells like `Script` would need a `Functor` instance.

Reply:
agree, will do it next time in cardano-addresses